usr/new/notes/man/nfadmin.8
.TH NFADMIN 8 "University of Illinois"
.SH NAME
nfadmin \- Change the status of a set of Notesfiles
.SH SYNOPSIS
.B nfadmin
flags
topic
[ topic ... ]
.br
.B nfadmin 
flags
-f listfile
.SH DESCRIPTION
.B Nfadmin
simplifies the task of displaying and setting director page options 
of many notesfiles at once.
.PP
The 
.I "flags"
specify the status to be changed.
.TP
-a+
Allow anonymous submissions.
.TP
-a-
Disallow anonymous submissions.
.TP
-n+
Allow network transmissions/receptions.
.TP
-n-
Disallow network transmissions/receptions.
.TP
-o+
Open the notesfile.  Allow local access by non-directors (subject to
access permissions).
.TP
-o-
Close the notesfile.  Disallow local access by non-directors.
.TP
-A+
Flag this notesfile as an archive notesfile.
(This does NOT mean that the notesfile will be archived.  You can't archive
an archive!)
.TP
-A-
Flag this notesfile as a non-archive notesfile.  This does not affect what
happens to notes when they expire.
.TP
-M+
Flag this notesfile as a moderated notesfile.  This causes commands that
create new notes or responses to create a mail message to the newsgroup
moderator instead.  A list of notesfiles and the mail address of the
moderator is maintained in /usr/spool/notes/.utilities/moderators.
.TP
-M-
Flag this notesfile as a non-moderated notesfile.
.TP
-L+
Flag this notesfile as a local notesfile.  Local notesfiles don't ask
whether to include a .signature file at the end of a new note or response.
.TP
-L-
Flag this notesfile as a non-local notesfile.  If the NFSIG environment
variable is set, notes will prompt whether or not to include the signature
file named by NFSIG at the end of a new note or response.
.TP
-e=NN
Set the expiration time of this notesfile to NN days.  (NN should be replaced
with a positive number).
.TP
-e=0
Set the expiration time of this notesfile to DEFAULT.  The number specified
on the nfarchive command line will be used to determine expiration time.
If no time is specified on the nfarchive command line, 14 days is used.
.TP
-e=-1
Set the expiration time of this notesfile to NEVER.
.TP
-W=0
Set the working set size of this notesfile to the command line default.
.TP
-W=NN
Set the working set size of this notesfile to NN.
.TP
-l=NNNNN
Set the maximum note size of an individual note in this notesfile to NNNNN.
(Warning:  negative numbers and numbers under 1000 are a very bad idea here.
The usual notes default of 100000 is too small for sources groups and
mod.map.  Please consider raising the limit for these groups to something
more reasonable, like 500000 or 2000000.  Also check your parms.h file;
notes has a "hard limit", and if you set the maximum size over the hard
limit, the hard limit still applies, even though the individual maximum
is still displayed).
.TP
-E=default
Set the expiration action of this notesfile to the command line default.
.TP
-E=delete
Set the expiration action of this notesfile to DELETE.
.TP
-E=archive
Set the expiration action of this notesfile to ARCHIVE
.TP
-D=default
Set the director message status of this notesfile to the command line default.
.TP
-D=on
Set the director message status of this notesfile to ON; only notes with
the director message ON will be removed by nfarchive.
.TP
-D=off
Set the director message status of this notesfile to OFF; only notes with
the director message OFF will be removed by nfarchive.
.TP
-D=nocare
Set the director message status of this notesfile to NOCARE; the director
message flag is not used by nfarchive in selecting notes for archive/deletion.
.TP
-f file
The file specified in the argument following the -f flag contains a series
of lines, each specifying a notesfile or group of notesfiles (wildcard) to 
be displayed or changed.
.PP
If no flags are specified (except possibly -f), no changes are made, but
the current state of the notesfiles is displayed.  If flag(s) are specified,
the new state after the changes is displayed, in the following format:
.sp
.nf
notesfile title NetW Open Anon Arch WSet Arch. Keep DirM   #     Max
                -------status------ Size Time  Actn Stat Notes   Size
net.bugs.2bsd : YES  YES   NO   NO     0   12  ARCH DFLT     3  500000
net.bugs.4bsd : YES  YES   NO   NO     2   12  ARCH DFLT    26  500000
net.bugs.usg  : YES  YES   NO   NO     0   12  ARCH DFLT     8  500000
net.bugs.uucp : YES  YES   NO   NO     0   12  ARCH DFLT     9  500000
net.bugs.v7   : YES  YES   NO   NO     2   12  ARCH DFLT     1  500000
nfmaint       : YES  YES   NO   NO     0 NEVER DEL  DFLT     3  500000
nfgripes      : YES  YES   NO   NO     0   14  DEL  DFLT     0  500000
net.micro.6809:  NO  YES   NO   NO     0    1  DEL  DFLT     0  500000
net.announce  : YES  YES   NO   NO     0 DFLT  ARCH DFLT     0  500000
.fi
(The output has been squished slightly to allow it to fit in the manual
page format).
.PP
All users are allowed to run this program.
The changes are only applied to notesfiles for which the executing
user is a director.
.PP
This program lives in the notesfile utility directory,
typically ``/usr/spool/notes/.utilities''.
.SH EXAMPLES
.in +1i
nfadmin -n+ -o+ -a- -e=14 "net.*"
.br
nfadmin -o- -f noteslist
.in
.PP
In the first example, all of the net.* notesfiles are set to networked,
open, non-anonymous, with an expiration time of 14 days.
The second example closes access to non-directors for all of the files
listed in the file "noteslist".  
.SH BUGS
.PP
More parameters ought to be changable with this program.
.PP
Nfadmin has not been tested with systems that can have notesfile names
longer than 14 characters.  The nfadmin code does not have such dependencies
except in the output format, but the columns may look ugly.  
.PP
Nfadmin will probably stop working when a certain news administrator
creates newsgroup names larger than any of the PDP-11 max data space 
(64k), the maximum argument/environment to exec (something around 10k), or 
the total amount of virtual memory on your machine.
.PP
.SH FILES
.PD 0
.TP 35
/usr/spool/notes
the default notesfile data base
.PD
.SH SEE ALSO
mknf(8),
notes(1),
nfcomment(3),
.br
.ul
The Notesfile Reference Manual
.SH AUTHORS
.nf
Gordon Burditt (sneaky!gordon)
Tandy Systems Design

created this program using the Notesfiles system by

Ray Essick (uiucdcs!essick, essick%uiuc@csnet-relay.arpa)
University of Illinois at Urbana-Champaign
Urbana, IL

as a base.
.fi


usr/new/notes/src/nfadmin.c
#define		MAINLINE
#include	"parms.h"
#include	"structs.h"

#ifdef	RCSIDENT
static char rcsid[] = "$Header: nfadmin.c,v 1.8 87/07/01 13:47:49 paul Exp $";
#endif	RCSIDENT

/*
 *	Any normal user can run this program. It only allows changes
 *	to notesfiles the user is a director in.
 *
 */


main (argc, argv)
int     argc;
char  **argv;
{
    struct io_f io;
    int     argn;
    struct nflist_f *nfptr;
    int	    change, newexp;
    int	    newanon, newopen, newnet, newarch, newlocal, newmod;
    int	    newwset, newdir, newact, newlen;

    startup (argc, argv);				/* common init code */

    newdir = newact = newwset = newlen = newexp = -2;
    newanon = newopen = newnet = newarch = newlocal = newmod = change = 0;
    for ( argc--, argv++; argc > 0 && argv[0][0] == '-'; argc--, argv++)
    {
	switch(argv[0][1])
	{
	case 'f':	
		readrc(argv[1]);
		argv++, argc--;
		break;
	case 'a':
		change = 1;
		newanon = newval(argv[0][2]);
		break;
	case 'n':
		change = 1;
		newnet = newval(argv[0][2]);
		break;
	case 'o':
		change = 1;
		newopen = newval(argv[0][2]);
		break;
	case 'L':
		change = 1;
		newlocal = newval(argv[0][2]);
		break;
	case 'M':
		change = 1;
		newmod = newval(argv[0][2]);
		break;
	case 'A':
		change = 1;
		newarch = newval(argv[0][2]);
		break;
	case 'e':
		change = 1;
		if (argv[0][2] == '=')
		{
			newexp = atoi(&argv[0][3]);
		} else usage();
		break;
	case 'W':
		change = 1;
		if (argv[0][2] == '=')
		{
			newwset = atoi(&argv[0][3]);
		} else usage();
		break;
	case 'l':
		change = 1;
		if (argv[0][2] == '=')
		{
			newlen = atoi(&argv[0][3]);
		} else usage();
		break;
	case 'E':
		change = 1;
		if (argv[0][2] == '=')
		{
			if (strncmp(&argv[0][3], "arch", 4) == 0)
				newact = KEEPYES;
			else if (strncmp(&argv[0][3], "del", 3) == 0)
				newact = KEEPNO;
			else if (strncmp(&argv[0][3], "def", 3) == 0)
				newact = KEEPDFLT;
			else if (strncmp(&argv[0][3], "dfl", 3) == 0)
				newact = KEEPDFLT;
			else usage();
		} else usage();
		break;
	case 'D':
		change = 1;
		if (argv[0][2] == '=')
		{
			if (strncmp(&argv[0][3], "on", 2) == 0)
				newdir = DIRON;
			else if (strncmp(&argv[0][3], "off", 3) == 0)
				newdir = DIROFF;
			else if (strncmp(&argv[0][3], "no", 2) == 0)
				newdir = DIRNOCARE;
			else if (strncmp(&argv[0][3], "anyon", 5) == 0)
				newdir = DIRANYON;
			else if (strncmp(&argv[0][3], "def", 3) == 0)
				newdir = DIRDFLT;
			else if (strncmp(&argv[0][3], "dfl", 3) == 0)
				newdir = DIRDFLT;
			else usage();
		} else usage();
		break;
	default:
		printf("Bad flag %c\n", argv[0][1]);
		usage();
		break;
	}
    }

    for (argn = 0; argn < argc; argn++)
    {
	expand (argv[argn]);				/* load it */
    }

    while ((nfptr = nextgroup ()) != (struct nflist_f *) NULL)
    {
	if (init (&io, nfptr -> nf_name) < 0)		/* open */
	{
	    printf ("%s:	couldn't open\n", nfptr -> nf_name);
	    continue;
	}
	if (change)
	{
	    if (!allow (&io, DRCTOK))			/* a director? */
	    {
		printf ("%s:	Not a director\n", nfptr -> nf_name);
		closenf (&io);
		continue;
	    }
	    locknf(&io, DSCRLOCK);
	    getdscr(&io, &io.descr);
	    if (newnet)
		if (newnet > 0)
			io.descr.d_stat |= NETWRKD;
		else
			io.descr.d_stat &= ~NETWRKD;
	    if (newanon)
		if (newanon > 0)
			io.descr.d_stat |= ANONOK;
		else
			io.descr.d_stat &= ~ANONOK;
	    if (newopen)
		if (newopen > 0)
			io.descr.d_stat |= OPEN;
		else
			io.descr.d_stat &= ~OPEN;
	    if (newlocal)
		if (newlocal > 0)
			io.descr.d_stat |= LOCAL;
		else
			io.descr.d_stat &= ~LOCAL;
	    if (newmod)
		if (newmod > 0)
			io.descr.d_stat |= MODERATED;
		else
			io.descr.d_stat &= ~MODERATED;
	    if (newarch)
		if (newarch > 0)
			io.descr.d_stat |= ISARCH;
		else
			io.descr.d_stat &= ~ISARCH;
	    if (newexp != -2)
		io.descr.d_archtime = newexp;
	    if (newwset != -2)
		io.descr.d_workset = newwset;
	    if (newlen != -2)
		io.descr.d_longnote = newlen;
	    if (newact != -2)
		io.descr.d_archkeep = newact;
	    if (newdir != -2)
		io.descr.d_dmesgstat = newdir;

	    putdscr(&io, &io.descr);
	    unlocknf(&io, DSCRLOCK);

	}
	notesum(&io.descr);
	closenf (&io);
    }
    exit (0);						/* all done */
}

usage (name)						/* how to invoke */
char   *name;
{
    fprintf (stderr,
	    "Usage: %s <flags> <notesfile> [<notesfile>...]\n",
	    "nfadmin");
    fprintf(stderr,
	    "-a+\tanonymous on\n-a-\tanonymous off\n-n+\tnetworked\n-n-\tnot networked\n-o+\topen\n-o-\tclosed\n-A+\tis archive\n-A-\tis not archive\n-e=n\texpiration time\n-e=-1\tnever expire\n");
    exit (1);
}

char *yes(b)
int	b;
{
	if (b)
		return "YES ";
	else	return " NO ";
}

int	flag = 0;

notesum(d)
struct descr_f *d;
{
/*
notesfile name  NetW Open Anon Arch   Wset archt kp  msgst    #note
xxxxxxxxxxxxxx: Y/N  Y/N  Y/N  Y/N   xxxxx xxxxx ARCH xxxx    xxxxx
*/

	char archtime[10], archkeep[10], msgstat[10];

	if (!flag)
	{
		printf("    notesfile title     NetW Open Anon Arch Locl Mod    WSet Arch. Keep DirM     #         Max\n");
		printf("                        ------------status-----------   Size Time  Actn Stat   Notes       Size\n");
	}
	flag = 1;
	switch((int) d->d_dmesgstat)
	{
	case DIRDFLT:	strcpy(msgstat, "DFLT");
			break;
	case DIRON:	strcpy(msgstat, " ON ");
			break;
	case DIROFF:	strcpy(msgstat, "OFF ");
			break;
	case DIRANYON:	strcpy(msgstat, "ANY ");
			break;
	case DIRNOCARE:	strcpy(msgstat, "IGNR");
			break;
	default:	strcpy(msgstat, "????");
			break;
	}
	switch((int) d->d_archkeep)
	{
	case KEEPDFLT:	strcpy(archkeep, "DFLT");
			break;
	case KEEPNO:	strcpy(archkeep, "DEL ");
			break;
	case KEEPYES:	strcpy(archkeep, "ARCH");
			break;
	default:	strcpy(archkeep, "????");
			break;
	}
	if (d->d_archtime == -1)
		strcpy(archtime, "NEVER");
	else if (d->d_archtime == 0)
		strcpy(archtime, "DFLT ");
	else	sprintf(archtime, "%4d ", d->d_archtime);


	printf("%22.22s: %s %s %s %s %s %s  %5d %s %s %s   %5d%12d\n",
		d->d_title, yes(d->d_stat & NETWRKD), yes(d->d_stat & OPEN),
		yes(d->d_stat & ANONOK), yes(d->d_stat & ISARCH), 
		yes(d->d_stat & LOCAL), yes(d->d_stat & MODERATED), 
		d->d_workset, archtime, archkeep, msgstat, d->d_nnote, 
		d->d_longnote);

}
int newval(c)
char	c;
{
	switch(c)
	{
	case '\0':
	case '+':
		return 1;
	case '-':
		return -1;
	}
	usage();
}
